### PR TITLE
Refactor/diagnostic agnostic

### PR DESCRIPTION
**Pull request by siwane**

### CODE REVIEW

File: controller/Diagnostic.php [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    protected $currentDelivery   = null;
+    protected $dataTable;
+
+    /**
+     * Display the list of all readiness checks performed on the given test center
+     * It also allows launching new ones.
+     */
+    public function index()
+    {
+        $diagnostics       = $this->getDiagnosticDataTable()->getDiagnostics($this->getRequestOptions());
```
Comment (hutnikau) [inline controller/Diagnostic.php]: Why you left so much spaces after `=` here and below after `=>`?

**Comment by hutnikau** [comment]
Please try to keep PRs as short as possible. Almost impossible to review and test such big PR properly.

**Comment by siwane** [comment]
Sorry for this large PR, goal was to integrate existing tool from taoProctoring

File: controller/Diagnostic.php [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+        foreach($data as $key => $value) {
+            if (is_array($value) || is_object($value)) {
+                $data[$key] = json_encode($value);
+            }
+        }
+
+        $this->defaultData();
+        $this->setData('userLabel', SessionManager::getSession()->getUserLabel());
+        $this->setData('clientConfigUrl', $this->getClientConfigUrl());
```
Comment (siwane) [inline controller/Diagnostic.php]: Done

File: controller/Diagnostic.php [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+     * Display the diagnostic runner
+     */
+    public function diagnostic()
+    {
+        $data = array(
+            'title'  => __('Readiness Check'),
+            'config' => $this->loadConfig(),
+        );
+
+        foreach($data as $key => $value) {
```
Comment (siwane) [inline controller/Diagnostic.php]: Done

File: controller/Diagnostic.php [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+     * Gets the data table request options
+     *
+     * @param array $defaults
+     * @return array
+     */
+    protected function getRequestOptions(array $defaults = []) {
+
+        $defaults = array_merge($this->getDefaultOptions(), $defaults);
+
+        $page = $this->hasRequestParameter('page') ? $this->getRequestParameter('page') : $defaults['page'];
```
Comment (siwane) [inline controller/Diagnostic.php]: Already discussed with JS, he will refactor this part to improve that and able to work with paginated storage

File: model/diagnostic/DiagnosticDataTable.php [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+use oat\taoClientDiagnostic\model\storage\PaginatedSqlStorage;
+use Zend\ServiceManager\ServiceLocatorAwareInterface;
+use Zend\ServiceManager\ServiceLocatorAwareTrait;
+
+/**
+ * This temporary helpers is a temporary way to return data to the controller.
+ * This helps isolating the mock code from the real controller one.
+ * It will be replaced by a real service afterward.
+ */
+class DiagnosticDataTable implements ServiceLocatorAwareInterface
```
Comment (siwane) [inline model/diagnostic/DiagnosticDataTable.php]: Already discussed with JS, he will refactor this part to improve that and able to work with paginated storage

File: model/diagnostic/DiagnosticDataTable.php [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+     * @return array
+     * @throws \common_exception_NoImplementation
+     */
+    public function getDiagnostics($options = array())
+    {
+        return $this->paginate($this->getStorage(), $options, function($data) {
+            foreach($data as $idx => $row) {
+                $rowData = [
+                    'id'          => $row[PaginatedSqlStorage::DIAGNOSTIC_ID],
+                    'workstation' => $row[PaginatedSqlStorage::DIAGNOSTIC_WORKSTATION] . ' (' . $row[PaginatedSqlStorage::DIAGNOSTIC_IP] . ')',
```
Comment (siwane) [inline model/diagnostic/DiagnosticDataTable.php]: Done, but strange that you had an error

File: model/diagnostic/DiagnosticDataTable.php [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+     *
+     * @return PaginatedStorage
+     * @throws StorageException
+     */
+    protected function getStorage()
+    {
+        if (! $this->storage) {
+            $storage = $this->getServiceLocator()->get(Storage::SERVICE_ID);
+            if (! $storage instanceof PaginatedStorage) {
+                throw new StorageException(__('The storage service provided to store the diagnostic results must be upgraded to support reads!'));
```
Comment (siwane) [inline model/diagnostic/DiagnosticDataTable.php]: yes

File: model/diagnostic/Paginator.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+namespace oat\taoClientDiagnostic\model\diagnostic;
+
+use oat\taoClientDiagnostic\model\storage\PaginatedStorage;
+
+/**
+ * Provides common data helper for datatable component.
+ */
+class Paginator
```
Comment (siwane) [inline model/diagnostic/Paginator.php]: yes, JS is refactoring the proctoring, all duplicate code will be merge in tao, close to datatable.
This part will be merge in tao

File: scripts/update/Updater.php [hunk 9]
```diff
@@ -324,5 +326,31 @@ public function update($initialVersion)
         }
 
         $this->skip('1.10.2', '1.13.2');
+
+        if ($this->isVersion('1.13.2')) {
+
+            $storageService  = $this->getServiceManager()->get(Storage::SERVICE_ID);
+
+            if ($storageService instanceof Sql && ! $storageService instanceof PaginatedStorage) {
```
Comment (siwane) [inline scripts/update/Updater.php]: You can use only compatibilityChecker with SQL storage. But if you want diagnostic list you need paginatedStorage but it's not required

**Pull request merged**